lib/model/migrations/20171010-01-auth.js

```diff
@@ -0,0 +1,79 @@
+
+const up = (knex, Promise) => {
+  const createActees = knex.schema.createTable('actees', (actees) => {
+    actees.string('id', 36).primary();
+    actees.string('species', 36);
+  });
+  const createActors = knex.schema.createTable('actors', (actors) => {
+    actors.increments('id');
+    actors.enu('type', [ 'system', 'user', 'group', 'proxy' ]).notNull();
+    actors.string('acteeId', 36).notNull();
+    actors.string('displayName', 64).notNull();
+    actors.jsonb('meta');
+
+    actors.dateTime('createdAt');
+    actors.dateTime('updatedAt');
+    actors.dateTime('deletedAt');
+
+    actors.foreign('acteeId').references('actees.id');
+
+    actors.index('type');
+  });
+  const createUsers = knex.schema.createTable('users', (users) => {
+    users.integer('actorId').primary();
+    users.string('password', 64);
+    users.string('mfaSecret', 20);
+    users.string('email', 320).unique().notNull();
+
+    users.dateTime('updatedAt');
+
+    users.foreign('actorId').references('actors.id');
+    users.index('email');
+  });
+  const createSessions = knex.schema.createTable('sessions', (sessions) => {
+    sessions.integer('actorId').notNull();
+    sessions.string('token', 64).notNull();
+    sessions.dateTime('expires').notNull();
+
+    sessions.dateTime('createdAt');
+
+    sessions.foreign('actorId').references('actors.id');
+    sessions.index([ 'actorId', 'expires' ]);
+  });
+  const createMemberships = knex.schema.createTable('memberships', (memberships) => {
```

Comment:
Rather than a single memberships table, what do you think about a separate table for each membership type? I think that's a pretty common pattern — the formhub logger implements it, for example — though of course that doesn't necessarily make it good. In the example of a campaign accessible to campaign users, other individual users, and other user groups, that would mean we'd have three tables for users, campaign users, and user group users. Would that be more or less performant than a recursive query?

Reply:
So, this model only thinks in the other direction. Users don't "belong" to a campaign; actors are granted rights on the campaign. The only canonical user group concept is the actor/membership table duo. Remember also that "users" are not a part of the permissioning system here. Users are each tied to an actor, and that actor is what we operate on. So there aren't users, user groups, and campaign users—there are only actors and grants.